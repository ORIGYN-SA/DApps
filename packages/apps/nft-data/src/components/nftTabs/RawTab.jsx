import React from 'react';
import { JsonToTable } from "react-json-to-table";

/* import useSite from '../../hooks/useSite'; */

const RawTab = ({metadata}) => {
  
    /* const {getMetadata, metadata} = useSite(); */
    /* useEffect(() => {
        console.log(metadata);
       }, [metadata]); */
      /* useEffect(() => {
        
        getMetadata();
        console.log(metadata);
      }, []); */
    return (
        <div>
{/*             <JSONViewer json={[metadata]}/> */}
            {/* {metadata} */}
            <JsonToTable json={metadata} />
            {/* <pre>{JSON.stringify(metadata, null, 2) }</pre> */}
           {/*  {JSON.stringify(metadata)} */}
        </div>
    )
}

export default RawTab;
