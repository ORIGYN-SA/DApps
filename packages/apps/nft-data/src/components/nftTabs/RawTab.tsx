import React from 'react';
import { JsonToTable } from 'react-json-to-table';

const RawTab = ({ metadata }: any) => {
  return (
    <div>
      <JsonToTable json={metadata} />
    </div>
  );
};

export default RawTab;
