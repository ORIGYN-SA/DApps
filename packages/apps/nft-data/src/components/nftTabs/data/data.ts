import { useContext } from 'react';
import { useRoute, AuthContext } from '@dapp/features-authentication';
import { getNft, getNftCollectionMeta, OrigynClient } from '@origyn/mintjs';

export type Nft_Data = {
  name: string;
  value: string | number;
  immutable: boolean;
  level: Level;
};

export type Level = 'Asset' | 'App' | 'Id' | 'Permission';

export type Permission = {
  level: Level;
  type: 'Permission' | 'Write' | 'Read';
  name: string;
  value?: string;
  list?: any[];
  immutable: boolean;
};

const getMetadata = async () => {
  const { actor } = useContext(AuthContext);
  const { tokenId, canisterId } = await useRoute();
  await OrigynClient.getInstance().init(true, canisterId, { actor });
  if (tokenId) {
    const response = await getNft(tokenId);
    console.log('responseNFT', response);
    const metadata = await response.ok.metadata.Class;
    console.log('metadataNft', metadata);
    return metadata;
  } else {
    console.log('no token id - getNftCollectionMeta');
    const response = await getNftCollectionMeta();
    console.log('responseCollection', response);
    const metadata = await response.ok.metadata[0].Class;
    console.log('collectionMeta', metadata);
    return metadata;
  }
};

export const getData = async () => {
  const { tokenId } = await useRoute();
  let Data_Array: Nft_Data[] = [];
  const Metadata = await getMetadata();
  console.log('Metadata', Metadata);
  if (Metadata) {
    try {
      Data_Array.push(
        {
          name: 'Id',
          value: await Metadata.filter((res) => {
            return res.name === 'id';
          })[0].value.Text,
          immutable: await Metadata.filter((res) => {
            return res.name === 'id';
          })[0].immutable,
          level: 'Id',
        },
        {
          name: 'Owner',
          value: await Metadata.filter((res) => {
            return res.name === 'owner';
          })[0].value.Text,
          immutable: await Metadata.filter((res) => {
            return res.name === 'owner';
          })[0].immutable,
          level: 'Id',
        },
        {
          name: 'App_Id',
          value: await Metadata.filter((res) => {
            return res.name === '__apps';
          })[0].value.Array.thawed[0].Class[0].value.Text,
          immutable: await Metadata.filter((res) => {
            return res.name === '__apps';
          })[0].value.Array.thawed[0].Class[0].immutable,
          level: 'App',
        },
        {
          name: 'Read',
          value: await Metadata.filter((res) => {
            return res.name === '__apps';
          })[0].value.Array.thawed[0].Class[1].value.Text,
          immutable: await Metadata.filter((res) => {
            return res.name === '__apps';
          })[0].value.Array.thawed[0].Class[1].immutable,
          level: 'Permission',
        },
      );
      // Assets level - Only for NFTs, Not for collecrion
      if (tokenId) {
        Data_Array.push(
          {
            name: 'Preview',
            value: await Metadata.filter((res) => {
              return res.name === 'preview_asset';
            })[0].value.Text,
            immutable: await Metadata.filter((res) => {
              return res.name === 'preview_asset';
            })[0].immutable,
            level: 'Asset',
          },
          {
            name: 'Hidden',
            value: await Metadata.filter((res) => {
              return res.name === 'hidden_asset';
            })[0].value.Text,
            immutable: await Metadata.filter((res) => {
              return res.name === 'hidden_asset';
            })[0].immutable,
            level: 'Asset',
          },
          {
            name: 'Primary',
            value: await Metadata.filter((res) => {
              return res.name === 'primary_asset';
            })[0].value.Text,
            immutable: await Metadata.filter((res) => {
              return res.name === 'primary_asset';
            })[0].immutable,
            level: 'Asset',
          },
          {
            name: 'Experience',
            value: await Metadata.filter((res) => {
              return res.name === 'experience_asset';
            })[0].value.Text,
            immutable: await Metadata.filter((res) => {
              return res.name === 'experience_asset';
            })[0].immutable,
            level: 'Asset',
          },
        );
      }
      // Apps level
      const apps = await Metadata.filter((res) => {
        return res.name === '__apps';
      })[0].value.Array.thawed[0].Class[4].value.Class;
      let i: string;
      for (i in apps) {
        console.log(apps[i]);
        if (apps[i].value.hasOwnProperty('Principal')) {
          Data_Array.push({
            name: apps[i].name,
            value: apps[i].value.Principal.toText(),
            immutable: apps[i].immutable,
            level: 'App',
          });
        } else if (apps[i].value.hasOwnProperty('Nat')) {
          Data_Array.push({
            name: apps[i].name,
            value: apps[i].value.Nat.toString(),
            immutable: apps[i].immutable,
            level: 'App',
          });
        } else {
          Data_Array.push({
            name: apps[i].name,
            value: apps[i].value.Text,
            immutable: apps[i].immutable,
            level: 'App',
          });
        }
      }
    } catch (e) {
      console.log(e);
    }
  }
  console.log('Data', Data_Array);
  return Data_Array;
};

export const getPermissions = async () => {
  let Allowed_Array: Permission[] = [];
  const Metadata = await getMetadata();

  if (Metadata) {
    try {
      let array_write = await Metadata.filter((res) => {
        return res.name === '__apps';
      })[0].value.Array.thawed[0].Class[2].value.Class;
      let i: any;
      for (i in array_write) {
        if (array_write[i].value.hasOwnProperty('Array')) {
          let list = [];
          let j: any;
          for (j in array_write[i].value.Array.thawed) {
            list.push(array_write[i].value.Array.thawed[j].Principal.toText());
          }

          Allowed_Array.push({
            level: 'Permission',
            name: array_write[i].name,
            immutable: array_write[i].immutable,
            list: list,
            type: 'Write',
          });
        } else {
          Allowed_Array.push({
            level: 'Permission',
            name: array_write[i].name,
            immutable: array_write[i].immutable,
            value: array_write[i].value.Text,
            type: 'Write',
          });
        }
      }

      let array_permission = await Metadata.filter((res) => {
        return res.name === '__apps';
      })[0].value.Array.thawed[0].Class[3].value.Class;
      let k: any;
      for (k in array_permission) {
        if (array_permission[k].value.hasOwnProperty('Array')) {
          let list = [];
          let l: any;
          for (l in array_permission[k].value.Array.thawed) {
            list.push(array_permission[k].value.Array.thawed[l].Principal.toText());
          }

          Allowed_Array.push({
            level: 'Permission',
            name: array_permission[k].name,
            immutable: array_permission[k].immutable,
            list: list,
            type: 'Permission',
          });
        } else {
          Allowed_Array.push({
            level: 'Permission',
            name: array_permission[k].name,
            immutable: array_permission[k].immutable,
            value: array_permission[k].value.Text,
            type: 'Permission',
          });
        }
      }
    } catch (e) {
      console.log(e);
    }
  }
  return Allowed_Array;
};
