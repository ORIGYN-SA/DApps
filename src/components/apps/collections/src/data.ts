// src/data.ts

interface Collection {
  name: string;
  checked: boolean;
  image: string;
  category_id: string;
  nftCount: number;
}

interface BackendResponse {
  collections: Collection[];
  totalPages: number;
}

const fetchCollectionsFromBackend = async (
  page: number,
  limit: number,
): Promise<BackendResponse> => {
  // Simulate more data for pagination
  const allCollections: Collection[] = [
    {
      name: 'Art',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Alice',
      nftCount: 120,
    },
    {
      name: 'Precious Metal',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Bob',
      nftCount: 85,
    },
    {
      name: 'Diamond',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Charlie',
      nftCount: 200,
    },
    {
      name: 'Jewelry',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Diana',
      nftCount: 150,
    },
    {
      name: 'Music',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Ethan',
      nftCount: 95,
    },
    {
      name: 'Real Estate',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Fiona',
      nftCount: 60,
    },
    {
      name: 'Photography',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Grace',
      nftCount: 110,
    },
    {
      name: 'Vintage Cars',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Henry',
      nftCount: 75,
    },
    {
      name: 'Antiques',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Irene',
      nftCount: 50,
    },
    {
      name: 'Collectible Coins',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Jack',
      nftCount: 130,
    },
    // Add more collections to have enough data for pagination
    {
      name: 'Sculptures',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Leo',
      nftCount: 90,
    },
    {
      name: 'Books',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Mia',
      nftCount: 65,
    },
    {
      name: 'Virtual Real Estate',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Nolan',
      nftCount: 80,
    },
    {
      name: 'Sports Memorabilia',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Olivia',
      nftCount: 115,
    },
    {
      name: 'Watches',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Peter',
      nftCount: 55,
    },
    {
      name: 'Wine Collection',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Quincy',
      nftCount: 70,
    },
    {
      name: 'Fashion',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Rachel',
      nftCount: 125,
    },
    {
      name: 'Handicrafts',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Sam',
      nftCount: 100,
    },
    {
      name: 'Rare Stamps',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Tina',
      nftCount: 40,
    },
    {
      name: 'Ceramics',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Uma',
      nftCount: 85,
    },
    {
      name: 'Modern Art',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Victor',
      nftCount: 95,
    },
    {
      name: 'Classic Cars',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Wendy',
      nftCount: 60,
    },
    {
      name: 'Historical Documents',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Xavier',
      nftCount: 30,
    },
    {
      name: 'Comics',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Zach',
      nftCount: 110,
    },
  ];

  const startIndex = (page - 1) * limit;
  const paginatedCollections = allCollections.slice(startIndex, startIndex + limit);
  const totalPages = Math.ceil(allCollections.length / limit);

  return {
    collections: paginatedCollections,
    totalPages,
  };
};
