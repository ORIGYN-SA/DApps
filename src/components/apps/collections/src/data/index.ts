import { Actor, HttpAgent } from '@dfinity/agent';
import { idlFactory } from './canisters/gold/did';
import { Principal } from '@dfinity/principal';
import { _SERVICE } from '../data/canisters/gold/interfaces/gld_nft';

export interface Collection {
  name: string;
  checked: boolean;
  image: string;
  creator: string;
  nftCount: number;
}

export interface BackendResponse {
  collections: Collection[];
  totalPages: number;
}

export const fetchCollectionsFromBackend = async (
  page: number,
  itemsPerPage: number,
): Promise<BackendResponse> => {
  try {
    const agent = new HttpAgent({ host: 'https://ic0.app' });
    const canisterId = 'io7gn-vyaaa-aaaak-qcbiq-cai';
    const actor = Actor.createActor<_SERVICE>(idlFactory, {
      agent,
      canisterId,
    });

    const logoResult = await actor.icrc7_logo();
    console.log('logoResult', logoResult);

    const collectionResult = await actor.collection_nft_origyn([]);
    if (!('ok' in collectionResult)) {
      throw new Error(
        `Erreur lors de la récupération de la collection : ${collectionResult.err.text}`,
      );
    }
    const collectionInfo = collectionResult.ok;

    if (!collectionInfo.token_ids || collectionInfo.token_ids.length === 0) {
      return { collections: [], totalPages: 0 };
    }

    const tokenIds = collectionInfo.token_ids[0];

    const nftResults = await actor.nft_batch_origyn(tokenIds);

    const collections: Collection[] = nftResults
      .map((nftResult, index) => {
        if ('ok' in nftResult) {
          const metadata = nftResult.ok.metadata;
          const name = extractName(metadata);
          const creator = extractCreator(metadata);
          const nftCount = 1; // TODO
          return {
            name,
            checked: true,
            image: logoResult ? logoResult : '',
            creator,
            nftCount,
          };
        } else {
          console.error(`Erreur pour le token ${tokenIds[index]}: ${nftResult.err.text}`);
          return undefined;
        }
      })
      .filter((item): item is Collection => item !== undefined);

    const totalPages = Math.ceil(collections.length / itemsPerPage);

    console.log('collections', collections);

    return {
      collections,
      totalPages,
    };
  } catch (error) {
    console.error('Erreur dans fetchCollectionsFromBackend:', error);
    return { collections: [], totalPages: 0 };
  }
};

const extractName = (metadata: any): string => {
  if ('Class' in metadata) {
    const properties = metadata.Class;

    const appsProp = properties.find((prop: any) => prop.name === '__apps');
    if (appsProp && 'Array' in appsProp.value) {
      const appsArray = appsProp.value.Array;

      for (const app of appsArray) {
        if ('Class' in app) {
          const appProps = app.Class;

          const appIdProp = appProps.find(
            (prop: any) =>
              prop.name === 'app_id' && prop.value.Text === 'com.origyn.metadata.general',
          );

          if (appIdProp) {
            const dataProp = appProps.find(
              (prop: any) => prop.name === 'data' && 'Class' in prop.value,
            );
            if (dataProp) {
              const dataProps = dataProp.value.Class;

              const displayNameProp = dataProps.find(
                (prop: any) => prop.name === 'display_name' && 'Text' in prop.value,
              );

              if (displayNameProp) {
                return displayNameProp.value.Text;
              }
            }
          }
        }
      }
    }
  }
  return 'Nom inconnu';
};

const extractCreator = (metadata: any): string => {
  if ('Class' in metadata) {
    const properties = metadata.Class;

    const ownerProp = properties.find(
      (prop: any) => prop.name === 'owner' && prop.value && prop.value.Principal,
    );

    if (ownerProp) {
      const principalValue = ownerProp.value.Principal;

      if (principalValue && typeof principalValue.toText === 'function') {
        return principalValue.toText();
      } else if (principalValue._arr) {
        const arrUnknown = Object.values(principalValue._arr);

        const arr = arrUnknown.map((item: unknown) => {
          if (typeof item === 'number') {
            return item;
          } else {
            return 0;
          }
        });

        const uint8Array = new Uint8Array(arr);
        const principal = Principal.fromUint8Array(uint8Array);
        return principal.toText();
      }
    }
  }
  return 'Créateur inconnu';
};
