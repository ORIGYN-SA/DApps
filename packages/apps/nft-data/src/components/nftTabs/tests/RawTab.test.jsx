import React from 'react';
import { render, screen, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom';
import data from './data';
import RawTab from '../RawTab';
describe('rawtab', () => {
    const { asFragment } = render(<RawTab metadata={data} />);
    it('should return rawtab rendered', () => {
        expect(screen.getByText('owner')).toBeInTheDocument();
        expect(screen.getByText('hidden_asset')).toBeInTheDocument();
        expect(screen.getByText('preview_asset')).toBeInTheDocument();
        expect(screen.getByText('primary_asset')).toBeInTheDocument();
        expect(screen.getByText('experience_asset')).toBeInTheDocument();
        expect(screen.getByText('id')).toBeInTheDocument();
        expect(screen.getByText('__apps')).toBeInTheDocument();
        expect(screen.getByText('library')).toBeInTheDocument();
       

    });
    it('should match snapshot', () => {
        expect(asFragment()).toMatchSnapshot();
    });

});