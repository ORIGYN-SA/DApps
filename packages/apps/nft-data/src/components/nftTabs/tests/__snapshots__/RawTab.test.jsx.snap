// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rawtab should match snapshot 1`] = `<DocumentFragment />`;
